### PR TITLE
Add a test for PreStop hooks.

### LINKED ISSUE #7091: PreStop in Lifecycle does not work
PreStop in Lifecycle is a very excited feature that I want it so much. With it, I can do some clean work before killing the Pod/Container, for example, I need this container stop receiving requests first, then complete processing its requests already received before it is killed. However It does not work as PostStop now. And I have checked the latest code and found it is still not implemented. Could anyone give a time point on it?

https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/kubelet/kubelet.go#L843

### PR DESCRIPTION
**Pull request by brendandburns**
Closes #7091

### CODE REVIEW

File: test/e2e/pre_stop.go [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+			Logf("Saw: %s", string(body))
+		}
+		return false, nil
+	})
+	expectNoError(err, fmt.Sprintf("validating pre-stop."))
+}
+
+var _ = Describe("PreStop", func() {
+	//This namespace is modified throughout the course of the test.
+	var namespace *api.Namespace
```
Comment (lavalamp) [inline test/e2e/pre_stop.go]: Please use the framework I just added-- ```f := NewFramework("prestop")```

File: test/e2e/pre_stop.go [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+
+	// At the end of the test, clean up by removing the pod.
+	defer func() {
+		By("deleting the server pod")
+		c.Pods(ns).Delete(podDescr.Name, nil)
+	}()
+
+	By("Waiting for pods to come up.")
+	err = waitForPodRunningInNamespace(c, podDescr.Name, ns)
+	expectNoError(err, fmt.Sprintf("waiting for server pod to start"))
```
Comment (quinton-hoole) [inline test/e2e/pre_stop.go]: Sprintf is unnecessary here. Either add pod name, or remove Sprintf.

File: test/e2e/pre_stop.go [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+
+	By(fmt.Sprintf("Creating tester pod %s in namespace %s", podDescr.Name, ns))
+	_, err = c.Pods(ns).Create(preStopDescr)
+	expectNoError(err, fmt.Sprintf("creating pod %s", preStopDescr.Name))
+	deletePreStop := true
+
+	// At the end of the test, clean up by removing the pod.
+	defer func() {
+		if deletePreStop {
+			By("deleting the tester pod")
```
Comment (quinton-hoole) [inline test/e2e/pre_stop.go]: By clauses should consistently use sentence case (first letter capitalized), as per https://github.com/golang/go/wiki/CodeReviewComments#error-strings

File: test/e2e/pre_stop.go [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+			c.Pods(ns).Delete(preStopDescr.Name, nil)
+		}
+	}()
+
+	err = waitForPodRunningInNamespace(c, preStopDescr.Name, ns)
+	expectNoError(err, fmt.Sprintf("waiting for tester pod to start"))
```
Comment (quinton-hoole) [inline test/e2e/pre_stop.go]: As above - either remove the unnecessary Sprintf, or include the pod name.

File: test/e2e/pre_stop.go [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+				return false, nil
+			}
+			if state.Received["prestop"] != 0 {
+				return true, nil
+			}
+			Logf("Saw: %s", string(body))
+		}
+		return false, nil
+	})
+	expectNoError(err, fmt.Sprintf("validating pre-stop."))
```
Comment (quinton-hoole) [inline test/e2e/pre_stop.go]: As above.

**Comment by quinton-hoole** [comment]
LGTM, once code review comments are addressed.

File: test/e2e/pre_stop.go (hunk 2 above)
Comment (brendandburns) [inline test/e2e/pre_stop.go]: done.

File: test/e2e/pre_stop.go (hunk 3 above)
Comment (brendandburns) [inline test/e2e/pre_stop.go]: done.

File: test/e2e/pre_stop.go (hunk 4 above)
Comment (brendandburns) [inline test/e2e/pre_stop.go]: done.

File: test/e2e/pre_stop.go (hunk 5 above)
Comment (brendandburns) [inline test/e2e/pre_stop.go]: done.

**Comment by brendandburns** [comment]
Comments addressed, please re-check.

Thanks!
--brendan

**Comment by quinton-hoole** [comment]
LGTM.  @brendandburns I assume that you've run this e2e test successfully?

**Comment by brendandburns** [comment]
@quinton-hoole yes, I've run it successfully several times.

**Comment by brendandburns** [comment]
@thockin (oncall) pingity ping ping

**Comment by quinton-hoole** [comment]
The merge queue got a bit jammed again today.  You have not been forgotten :-)

**Pull request merged**